### ISSUE TITLE
Pyright looses literal information on loops/generators?

### ISSUE THREAD

**Issue by fsouza** [microsoft/pyright#8035]
**Describe the bug**

I have a piece of code that contains a `dict[Literal[...], ...]`. When constructing values off that (from `.items()`), we lose information about the literal and end-up with a string. As far as I can tell, this only happens on loops/generators.

We can workaround the issue by hinting the type.

**Code or Screenshots**
Unfortunately the code where I ran into this issue is not open source. This is my attempt at creating a simple reproduction:

```python
from typing import Literal

X = Literal["x", "y", "z"]

all_items: dict[X, int] = {}

# first example: just grab the first (k, v) tuple, reveal_type reports expected type
kv0 = next(iter(all_items.items()))
reveal_type(kv0)

# second example, grab the first (k, v) via a generator, reveal_type reports str instead of the literal
kv1 = next((k, v) for k, v in all_items.items())
reveal_type(kv1)

# third example: same as above, but with explicit type annotation
kv2: tuple[X, int] = next((k, v) for k, v in all_items.items())
reveal_type(kv2)
```

Running pyright 1.1.365 on that file, I get the following output:

```
％ pyright a.py
/private/tmp/a.py
  /private/tmp/a.py:7:13 - information: Type of "kv0" is "tuple[Literal['x', 'y', 'z'], int]"
  /private/tmp/a.py:10:13 - information: Type of "kv1" is "tuple[str, int]"
  /private/tmp/a.py:13:13 - information: Type of "kv2" is "tuple[Literal['x', 'y', 'z'], int]"
0 errors, 0 warnings, 3 informations
```

With 1.1.364, I get:

```
％ pyright a.py
/private/tmp/a.py
  /private/tmp/a.py:7:13 - information: Type of "kv0" is "tuple[Literal['x', 'y', 'z'], int]"
  /private/tmp/a.py:10:13 - information: Type of "kv1" is "tuple[Literal['x', 'y', 'z'], int]"
  /private/tmp/a.py:13:13 - information: Type of "kv2" is "tuple[Literal['x', 'y', 'z'], int]"
0 errors, 0 warnings, 3 informations
```

With mypy 1.10.0, I get:

```
％ mypy a.py
a.py:7: note: Revealed type is "tuple[Union[Literal['x'], Literal['y'], Literal['z']], builtins.int]"
a.py:10: note: Revealed type is "tuple[Union[Literal['x'], Literal['y'], Literal['z']], builtins.int]"
a.py:13: note: Revealed type is "tuple[Union[Literal['x'], Literal['y'], Literal['z']], builtins.int]"
Success: no issues found in 1 source file
```

**VS Code extension or command-line**

First noticed it while running pyright as a language server in neovim, but can repro from the command line.

**Comment by erictraut** [comment]
Prior to 1.1.365, type inference for tuple expressions always retained literals. For example, `(1, 2)` was always inferred as `tuple[Literal[1], Literal[2]]`.

After significant feedback and debate, we decided to change this behavior to drop literals when a tuple expression is nested within a list, set, or dict expression. Prior to this change, the type of `[(1, ), (2, )]` was inferred as `list[tuple[Literal[1]] | tuple[Literal[2]]]`, but most users expected it to be `list[tuple[int]]`. The old behavior was especially problematic for list, set and dict types because all of them are invariant, so `list[tuple[Literal[1]] | tuple[Literal[2]]]` is not a subtype of `list[tuple[int]]`.

In your example for `kv1`, the tuple expression `(k, v)` is used within a comprehension, but it's part of a generator, not a list. Generators are covariant in their types, so I think it's reasonable to retain the pre-1.1.365 behavior in this case.

This change will be included in the next release.

**Comment by fsouza** [comment]
Thanks for the quick fix! Should that be the case with non-tuples too?

Another example we have is something like:

```python
X: Literal[...]

def get_thing(...) -> X: ...

xs = (get_thing(item) for item in some_input())
```

xs ends up being `Generator[str, None, None]` insted of `Generator[X, None, None]` (unless we annotate it explicitly)

(Let me know if if I should open another issue for this, as this behavior appears to have changed in 1.1.357, not in the last release)

**Comment by erictraut** [comment]
This change was made in 1.1.357 to address this issue: https://github.com/microsoft/pyright/issues/7586. I'm reluctant to undo this change.

And now that you point this out, I'm second guessing my change above regarding generator expressions. Maybe it would be best (for consistency) to never retain literals when inferring comprehensions unless bidirectional type inference indicates that the expected type includes a literal. That is, it might be best to stick with the behavior in 1.1.365 . To retain literals, you can provide an explicit type declaration like you've done with `kv2` above.

Thoughts?

**Comment by fsouza** [comment]
Hmmm I think the consistency argument is fair, we can add the explicit type definitions on our side.

**Comment by erictraut** [comment]
OK, I'm going to revert the change I made above and close this issue.
